pkg/build/api/types.go

```diff
@@ -228,8 +228,91 @@ type BuildStatus struct {
 
 	// Output describes the Docker image the build has produced.
 	Output BuildStatusOutput
+
+	// Stages contains details about each build stage.
+	Stages Stages
+}
+
+// Stages is a slice of build stages that have been recorded.
+type Stages []StageInfo
+
+// StageInfo contains details about a build stage.
+type StageInfo struct {
+	// StageName is the identifier for each build stage.
+	StageName StageName
+
+	// StartTime identifies when this stage started.
+	StartTime unversioned.Time
+
+	// EndTime identifies when this stage ended.
+	EndTime unversioned.Time
+
+	// Duration identifies how long this stage ran.
+	Duration unversioned.Duration
+
+	// Steps contains details about each build step within a build stage.
+	Steps Steps
+}
+
+// StageName is the identifier for each build stage.
+type StageName string
+
+// Valid StageNames
+const (
+	// StageFetch fetches the source. 
+	StageFetch StageName = "Fetch"
+
+	// StagePullImages pulls the docker images.
+	StagePullImages StageName = "PullImages"
+
+	// StageBuild builds the source.
+	StageBuild StageName = "Build"
+
+	// StagePost executes post steps.
+	StagePost StageName = "Post"
+
+	// StagePushImages pushes the docker images.
+	StagePushImages StageName = "PushImages"
+)
+
+// Steps is a slice of build steps that have been recorded within each build stage.
+type Steps []StepInfo
+
+// StepInfo contains details about a build step.
+type StepInfo struct {
+	// StepName is the identifier for each build step.
+	StepName StepName
+
+	// StartTime identifies when this step started.
+	StartTime unversioned.Time
+
+	// EndTime identifies when this step completed.
+	EndTime unversioned.Time
+
+	// Duration identifies how long this step ran.
+	Duration unversioned.Duration
 }
 
+// StepName is the identifier for each build step.
+type StepName string
+
+const (
+	// StepExecPostCommitHook executes the buildconfigs post commit hook.
+	StepExecPostCommitHook StepName = "ExecPostCommitHook"
+
+	// StepFetchSource fetches the source code.
+	StepFetchSource StepName = "FetchSource"
+
+	// StepDockerPullImage pulls the docker image.
+	StepDockerPullImage StepName = "DockerPullImage"
```

Comment:
should have different step names for the different images we pull (builder image, base image, input image, etc)